Support configuration services: inline objects

uncleDecart opened the pull request:
This is PR has two main parts:

- It refactors zedrouter's metadata server and switch from http mux to go-chi router
- It adds first route to send `description.json` file to metadata server
- It adds parsing  and publishing of `PatchEnvelope` structure to zedagent

It's still work in progress, but I'd be glad to get some feedback :) 

It has not yet been tested with controller, will provide updates here

**Edit**:
- `persistcache` library is now part of EVE
- https://github.com/lf-edge/eve-api/pull/12 is merged
- Downloading route is added

uncleDecart commented on pkg/pillar/cmd/zedrouter/metadata.go:
```diff
@@ -8,84 +8,13 @@
-	"path"
-	"strings"
 
+	"github.com/go-chi/chi/v5"
```
at some point gorilla was deprecated and had no active maintaners, I heard they found somebody, but better be safe than sorry, plus switching to go-chi costs us nothing since it is based on http.Mux and since we don't have tests for older routes, but it would be nice to have some kind of modern router (because we don't want to parse URL to get route parameters such as /api/v1/{patch_envelope_id}/{file_path}

christoph-zededa replied:
Okay, fine, can you then please remove `gorilla/mux` from `go.{mod,sum}`?

uncleDecart replied:
We do use it [wstunnelclient](https://github.com/lf-edge/eve/blob/master/pkg/pillar/zedcloud/wstunnelclient.go) so it might be a good conversation to think should we switch to gorilla and how we should switch gorilla websockets

christoph-zededa replied:
IMO we should not use both, because then we basically double the risk that one project gets deprecated and we have to fix it and you basically said "better be safe than sorry".

(Ideally we abstract it away from the underlying muxer so we have to change it only at one place, but that's too much for now, I think)

eriknordmark replied:
@uncleDecart Do we need anything fancy from the router necessating switching from the basic and standard http.NewServeMux?

uncleDecart replied:
Reason why I'm bringing go-chi here is because I want to use path parameters which, i.e. 
`/eve/v1/patch/{patch-id}/{file-name}`
It is not yet it PR, I'm implementing logic. We, of course can write our own parser, but I'm trying to avoid reinventing wheel here. Another reason I'm using go-chi is because it's 100% compatible with `http/net` and has no other dependencies except this package.

eriknordmark replied:
Makes sense.

uncleDecart replied:
So @christoph-zededa we do not use gorilla mux explicitly in EVE, however it's indirect dependency from gorilla ws, which is not EOL and also used to communicate with controller and in controller we are still using gorilla ws, so I suggest we keep using gorilla ws for now

christoph-zededa commented on pkg/pillar/pubsub/util.go:
```diff
@@ -15,6 +15,15 @@ import (
+// Note why json marshalling is used:
+// Type casting and associated type assertions in golang are only
+// useful for atoms in the type system. Hence you can't do a type
+// assertion and cast fo a struct of internal fields. This coupled
```
```suggestion
// assertion and cast of a struct of internal fields. This coupled
```

christoph-zededa commented on pkg/pillar/cmd/zedagent/parseconfig_test.go:
```diff
@@ -1254,3 +1264,73 @@ func TestParseSRIOV(t *testing.T) {
+func TestParsePatchEnvelope(t *testing.T) {
+	g := NewGomegaWithT(t)
+	getconfigCtx := initGetConfigCtx(g)
+
```
```suggestion
t.Parallel()
```
I am wondering why yetus is not complaining ...

Is gomega doing this under the hood?

uncleDecart replied:
AFAIK gomega is not testing framework, ginko is. So I'm not sure it's doing this under the hood...

milan-zededa commented on pkg/pillar/types/patchenvelopestypes.go:
```diff
@@ -0,0 +1,87 @@
+	BinaryBlobs []BinaryBlob `json:"binary-blobs"`
+}
+
+func GetZipArchive(root string, pe PatchEnvelopeInfo) (string, error) {
```
GetZipArchive should go into `pillar/utils`, `types` package should preferably contain only global type declarations and their methods.

uncleDecart replied:
We can change signature to :D 
```go
func (pe PatchEnvelopeInfo) GetZipArchive(root string) (string, error) {
//...
}
```

eriknordmark replied:
FWIW changing the signature makes sense; in any case this operators on a PE.

milan-zededa replied:
@uncleDecart It is not about the signature. Package `types` should only contain structures defining APIs between pillar microservices. This is imported by every microservice, including "mmagent" in the future. Having this method here will add dependencies and increase binary size (which is unnecessary for everything outside of zedbox, i.e. edgeview, mmagent, etc.).

uncleDecart replied:
Moved this code to utils

milan-zededa commented on pkg/pillar/cmd/zedrouter/zedrouter.go:
```diff
@@ -776,6 +781,20 @@ func (z *zedrouter) initSubscriptions() (err error) {
+		AgentName:   "zedagent",
+		MyAgentName: agentName,
+		TopicImpl:   types.PatchEnvelopes{},
+		Activate:    false,
```
Isn't it necessary to activate the subscription eventually?

uncleDecart replied:
Done

milan-zededa commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,31 @@
+
+## Use-case
+
+This package implements a persistance cache. The main use-case is to store data on filesystem and
```
Would be good to mention that the cache has key-value semantics.

uncleDecart replied:
Fair :)

uncleDecart replied:
Updated

uncleDecart commented on pkg/pillar/cmd/zedagent/parseconfig_test.go:
```diff
@@ -1254,3 +1264,73 @@ func TestParseSRIOV(t *testing.T) {
+	}
+
+	path, _ := os.Getwd()
+	persistCacheFolder := filepath.Join(path, "testPersist/")
```
Makes sense

uncleDecart replied:
Done

uncleDecart commented on pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:
```diff
@@ -0,0 +1,128 @@
+	"github.com/lf-edge/eve/pkg/pillar/types"
+)
+
+const PERSIST_CACHE_FILEPATH = "/persist/cache/"
```
Will change it

uncleDecart replied:
Moved it to type constants, done

uncleDecart commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,31 @@
+it is loads values from cache if they are present or created folder if there was none.
+`persistCache` structure has 3 operations:
+
+- *Get* value from in-memory cache
```
Will add this

uncleDecart replied:
updated

uncleDecart commented on pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:
```diff
@@ -0,0 +1,128 @@
+	}
+
+	metadata := artifact.GetBase64MetaData()
+	data := artifact.GetBase64Data()
```
We have two possible entities as patchEnvelopes: inline objects and externally referenced objects. Inline objects we receive as part of `EdgeDevConfig` and they are limited to 64Kbytes.
If there's anything more than 64Kbytes it should be created as externally referenced object. That is simply Volume referenced in patchEnvelope

eriknordmark replied:
@uncleDecart I think the whole EdgeDevConfig is limited to about 64Kbytes so if we want to support that size for each inlined object then I think we will have an issue? Makes sense to check the size limit for the EdgeDevConfig (if any) in eden and the Zededa commercial controller.

eriknordmark replied:
I wouldn't be a bad idea to have the Put function (or a variant of it) take a bytes.Buffer so we don't have to read lots of 64kbyte items into memory merely to write them to a file.

uncleDecart replied:
@eriknordmark, checked size on eden and Zededa commercial controller, we don't have any 64Kbytes restriction on EdgeDevConfig

eriknordmark replied:
Thanks

uncleDecart commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,184 @@
+
+type persistCache struct {
+	sync.Mutex
+	cache map[string]string
```
Inline patch envelope we get from `EdgeDevConfig` should be stored somewhere and then served directly to client. We just need to keep track of all files created, deleted, modified. Cache in this case is not that important.

eriknordmark replied:
A cache makes sense, but we don't need to read data into memory if you make Get() return an io.Reader (and make Put take an io.Reader). Then you can do an io.Copy  which will use a limited size buffer.

uncleDecart replied:
if we don't use WriteRename in Put, because there we will have to convert to slice of bytes which will load whole object in memory

uncleDecart replied:
@eriknordmark using Reader also implies that you read only once. And passing by value does not copy whole stream, I.e. 

```
type MyFavStruct struct {
  Field map[string]io.Reader
}

func main() {

s := strings.NewReader( "bazinga" )

a := MyFavStruct{
  Field: make(map[string]io.Reader)
}

a.Field["bobo"] = s
sr1 := new(strings.Builder)

io.Copy(sr1, s)
fmt.PrintIn(sr1.String())

sr2 := new(strings.Builder)
io.Copy(sr2, a.Field["bobo"])


fmt.Printin(sr2.String())

}

```

Will output bazinga and empty line 

I think we should stay with []byte or wrap it around bytes buffer for chunking.

Current implementation with cache and its usage won't load more than one object to byte slice.

eriknordmark replied:
Ok

uncleDecart commented on pkg/pillar/cmd/zedrouter/metadata_handlers.go:
```diff
@@ -0,0 +1,625 @@
+			w.Write(b)
+			return
+		} else {
+			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
```
done

uncleDecart commented:
Pushed new commits addressing some of the conversations in this PR and also adding support for external patch envelope objects

eriknordmark commented on pkg/pillar/cmd/zedrouter/pubsubhandlers.go:
```diff
@@ -634,3 +634,27 @@ func (z *zedrouter) handleAppInstDelete(ctxArg interface{}, key string,
 	z.unpublishAppInstMetadata(appInstMetadata)
 	z.log.Functionf("handleAppInstDelete(%s) done", key)
 }
+
+func (z *zedrouter) handlePatchEnvelopeCreate(ctxArg interface{}, key string,
+	configArg interface{}) {
+}
+
+func (z *zedrouter) handlePatchEnvelopeModify(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+}
+
+func (z *zedrouter) handlePatchEnvelopeDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+}
+
+func (z *zedrouter) handleVolumeStatusCreate(ctxArg interface{}, key string,
+	statusArg interface{}) {
+}
+
+func (z *zedrouter) handleVolumeStatusModify(ctxArg interface{}, key string,
+	statusArg interface{}, oldStatusArg interface{}) {
+}
+
+func (z *zedrouter) handleVolumeStatusDelete(ctxArg interface{}, key string,
+	statusArg interface{}) {
+}
```
Are there functions supposed to be empty? If so you don't need to register functions in NewSubscription().

uncleDecart replied:
Those functions are needed for external patch envelopes, removed declarations from this PR, will introduce them in separate PR

eriknordmark replied:
ok

eriknordmark commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,36 @@
+
+This package implements a persistance cache. The main use-case is to store data on filesystem and
+provide access to it via API. Limitation of current implementation is that data should be represented
+as string. This was developed for configuration services feature in EVE. Since there is an option for 
```
string implies valid UTF-8. Can't you define it to be a byte slice instead? That is much more general.

uncleDecart replied:
Makes sense I changed it to byte slice

eriknordmark commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,36 @@
+## API & Usage
+
+Main structure is `pesistCache` outside of `persistcache` package it could be created via `Load` function:
+it is loads values from cache if they are present or created folder if there was none.
```
Hard to parse. Is there a type about should say "create folder"?

uncleDecart replied:
Rephrased it

eriknordmark commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,36 @@
+## Design decisions
+
+### Why we are storing separate files and not saving whole structure as file?
++ If objects stored are large it takes less time to save/update them and less code
```
Presumably it also avoids a Put of one key from affecting the storage of another key, which could happen if it is a single file and the device is powered off before everything has been sync'ed to disk.

uncleDecart replied:
added it

eriknordmark commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,188 @@
+	pc.Lock()
+	defer pc.Unlock()
+
+	if _, err := os.Stat(filepath.Join(pc.tempDir, LOCK_FILE_NAME)); os.IsNotExist(err) {
```
Why do we need a lock file? Raises complexity  around cleaning up after a reboot and there is still a race between a Stat and a Create call.

With WriteRename as the underlying mechanism for modifying the files then even if multiple users (processes, goroutines) use persistcache, a Get of a key will either fetch the content for that key before a Put or after but not mixed/partial content. That should be sufficient AFAIU.

uncleDecart replied:
Yes, you're right, let's keep it simple since WriteRename will be enough

eriknordmark commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func isValidKey(key string) bool {
+	if strings.Contains(key, "/") {
```
What about names like ".." and "."?
I think it makes sense to call path.Clean() on the key (since it will replace .. with ../ etc).
Then you can check for "/".

uncleDecart replied:
Makes sense, added it

eriknordmark commented on pkg/pillar/persistcache/persistcache_test.go:
```diff
@@ -0,0 +1,140 @@
+	"testing"
+
+	"github.com/lf-edge/eve/pkg/pillar/persistcache"
+	"gotest.tools/assert"
```
Is there something missing in the test package we use in all the other unit tests?

uncleDecart replied:
No, I removed assert completely and used gomega

uncleDecart commented:
Renaming it back to inline objects, will add external objects and statuses in different PRs. Updated commits to include only code for inline binary artifacts, added documentation for Patch Envelopes and updated code from PR suggestions. I think we can merge this PR if there are no more comments or remarks

uncleDecart commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,188 @@
+package persistcache
```
Added it

eriknordmark replied:
Did you push that change? Doesn't show in github.

uncleDecart replied:
Now you should see it :)

uncleDecart commented:
Updated documentation, fixed some of the yetus warnings

eriknordmark commented on pkg/pillar/cmd/zedrouter/metadata_handlers.go:
```diff
@@ -0,0 +1,693 @@
+	w.Write([]byte(fmt.Sprintf("{\"message\": \"%s\"}", msg)))
+}
+
+func HandlePatchDownload(z *zedrouter) func(http.ResponseWriter, *http.Request) {
```
Still some revive and codespell issues from yetus to fix.

uncleDecart replied:
Updated

eriknordmark replied:
But still more below.

uncleDecart replied:
:smiling_face_with_tear:

uncleDecart replied:
Let's see what yetus have to say now

eriknordmark commented on pkg/pillar/types/patchenvelopestypes_test.go:
```diff
@@ -0,0 +1,113 @@
+package types
```
Also missing Copyright and License lines.

uncleDecart replied:
Now you can see it :)

eriknordmark replied:
yes.

uncleDecart commented on pkg/pillar/persistcache/README.md:
```diff
@@ -0,0 +1,31 @@
+This package implements a persistance cache. The main use-case is to store data on filesystem and
+provide access to it via API. Limitation of current implementation is that data should be represented
+as string. This was developed for configuration services feature in EVE. Since there is an option for 
+user to provide object to store via inline query, EVE has to store it even after reboot.
```
Added doc for it

uncleDecart commented:
Struggle goes on, fixed yetus issues and updates license files.

In addition, now inline objects are cleared every time EdgeDevConfig gets created/recreated 

Have yet to implement io.Reader for persistcache

eriknordmark reviewed (commented):
> This, of course create a possible state where app instance tries
to access file which is not there anymore and new config has not
been propagated yet to zedrouter, but in that case metadata server
will return error and app instance can either retry or request
description.json once again

What happens to the description.json we serve to the app instance as we do these delete and recreate? Could it still list the files but a fetch of those files might fail? Or will it exclude the files which are deleted until they have been recreated?
I don't know how likely app developers and other tools are to trip on this.
Couldn't we have the running EVE keep the sha256 of each inline object in memory and compare the patch envelopes against that, and only do the delete+recreate if a particular inline object has changed?

eriknordmark commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,146 @@
+	root  string
+}
+
+const FileMask = 0755
```
Do you need this to be exported?

uncleDecart replied:
No, I changed it

eriknordmark commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,146 @@
+}
+
+const FileMask = 0755
+const LockFileName = "persistcache.lock"
```
I thought you removed the lock file, but it is still here and in the test

uncleDecart replied:
removed it all

eriknordmark commented on pkg/pillar/cmd/zedrouter/zedrouter.go:
```diff
@@ -776,6 +796,33 @@ func (z *zedrouter) initSubscriptions() (err error) {
+		return err
+	}
+
+	// Infromation about volumes reffered in external patch envelopes
```
More yetus

uncleDecart replied:
fixed

uncleDecart commented:
Fighting with yetus yet again :) 

Addressed comments added

uncleDecart commented on pkg/pillar/persistcache/persistcache.go:
```diff
@@ -0,0 +1,146 @@
+const FileMask = 0755
+const LockFileName = "persistcache.lock"
+
+type InvalidKeyError struct{}
```
fixed

gkodali-zededa commented on pkg/pillar/cmd/zedrouter/metadata_handlers.go:
```diff
@@ -0,0 +1,713 @@
+			pe := peObj.(types.PatchEnvelopes)
+
+			remoteIP := net.ParseIP(strings.Split(r.RemoteAddr, ":")[0])
+			anStatus := z.lookupAppNetworkStatusByAppIP(remoteIP)
```
Do you check if the given app instance (caller of API) is part of the allowed list before reaching here?

uncleDecart replied:
I follow same logic as other handlers in resolving remoteAddr to uuid, then check if there are any patch envelopes available for given app instance uuid and return them

gkodali-zededa commented on pkg/pillar/cmd/zedagent/parsepatchenvelopes.go:
```diff
@@ -0,0 +1,153 @@
+	var blobsAfter []string
+	patchEnvelopes := config.GetPatchEnvelopes()
+	result := types.PatchEnvelopes{}
+	for _, pe := range patchEnvelopes {
```
Do we destroy and re-create patch envelopes every time configuration gets pulled from zedcloud? Does it happen even when the patch envelope did not change?

uncleDecart replied:
We destroy only inline objects, which are not referenced anymore when new configuration comes. If inline patch envelope did not change we will not override it.

uncleDecart commented on docs/ECO-METADATA.md:
```diff
@@ -311,3 +311,54 @@ EVE is generating diagnostic output on a console (if there is one) which summari
+
+    {
+        "AllowedApps":[
+            "17daa0ff-39d6-42be-a537-44c974276aec"
```
Updated

uncleDecart commented:
Addressed review comments

eriknordmark commented on docs/ECO-METADATA.md:
```diff
@@ -311,3 +311,51 @@ EVE is generating diagnostic output on a console (if there is one) which summari
+        "BinaryBlobs":[
+            {
+                "file-name":"textfile1.txt",
+                "file-sha":"e5f3f80da0f9a3add3540a80d214069079f71e879b87c51c4c3a258989294812",
```
Is there some comment to make detsecrets shut up for this line?

uncleDecart commented:
@eriknordmark I think that this PR is ready to be merged (there are some yetus errors still, but I don't think we should address them) let me know what you think

eriknordmark reviewed (approved):
LGTM but still plenty of yetus issues to fix.

uncleDecart commented:
@eriknordmark all checks are green, I added ignoring of high entropy strings for tests and removed sha from documentation examples, rest I fixed
